docs/developer/best_practices/commit.rst

```diff
@@ -0,0 +1,356 @@
+Commit best practices
+---------------------
+
+How to split a change into commits
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The golden rule to create good commits is to ensure that there is only one
+logical change per commit.
+
+This has several advantages amongst whose:
+
+- small commits are easier to review (a large pull request correctly divided
+  into logical commits is easier/faster to review than a medium-sized one with
+  less thought-out division)
+
+- logical commits are easier to revert (`e866b01f0553 <https://github.com/scality/metalk8s/commit/e866b01f05535925e80da20aca00417904422433>`_/`8208a170ac66 <https://github.com/scality/metalk8s/commit/8208a170ac66912ace018bcd00c058ad214d169b>`_)/cherry-pick
+  (`Pull request #1641 <https://github.com/scality/metalk8s/pull/1641>`_)
+
+- when looking for a regression (e.g. using ``git bisect``) it is easier to
+  find the root cause
+
+- make ``git log`` and ``git blame`` way more useful
+
+Examples
+^^^^^^^^
+
+Cosmetic changes
+::::::::::::::::
+
+Use a dedicated commit when you want to make cosmetic changes to the code
+(linting, whitespaces, alignment, renaming, …).
+
+Mixing cosmetics and functional changes is bad because the cosmetics (which
+tend to generate a lot of diff/noise) will obscure the important functional
+changes, making it harder to correctly determine whether the change is correct
+during the review.
+
+Example (Pull request `#1620 <https://github.com/scality/metalk8s/issues/1620>`_):
+
+- one commit for the cosmetic changes: `766f572e462c6933c8168a629ed4f479bb68a803 <https://github.com/scality/metalk8s/commit/766f572e462c6933c8168a629ed4f479bb68a803>`_
+
+- one commit for the functional changes: `3367fabdefc0b35d34bf7cf2fb0d33ff81f9fd5a <https://github.com/scality/metalk8s/commit/3367fabdefc0b35d34bf7cf2fb0d33ff81f9fd5a>`_
+
+Refactoring
+:::::::::::
+
+When introducing new features, you often have to add new helpers or refactor
+existing code. In such case, instead of having single commit with everything
+inside, you can either:
+
+1. first add a new helper: `29f49cbe9dfa <https://github.com/scality/metalk8s/commit/29f49cbe9dfa0b824c818d25d4a2f6965351e65d>`_
```

Comment:
I disagree with this one and don't consider it a 'good' commit. It adds utility code which is not used anywhere.

This breaks the `git revert` rule.

Instead, the commit adding the utility function could've also updated code to *use* this utility, if any. If no existing code can use it, then introduce it *with* the 'feature' code that *will*, otherwise the `git revert`/`git cherry-pick` rule doesn't apply again.

Reply:
The function to check if an image exists is not tied to the `uptodate` one.

I mean, yes `uptodate` does need this helper (and thus if you want to cherry-pick this "feature" you'll have to cherry-pick the helper as well), but the helper could also be used in another independent context (pre-check that some images exists, …) and thus could be cherry-picked independently.
Likewise, you may want to revert the `uptodate` stuff while keeping the helper around for another use.